# Credit Risk Resampling Techniques

In [29]:
import warnings
warnings.filterwarnings('ignore')

In [30]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

# Read the CSV into DataFrame

In [71]:
# Load the data
file_path = Path('Data\lending_data.csv')
df = pd.read_csv(file_path)
df.head()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672       own            52800        0.431818   
1     8400.0          6.692       own            43600        0.311927   
2     9000.0          6.963      rent            46100        0.349241   
3    10700.0          7.664       own            52700        0.430740   
4    10800.0          7.698  mortgage            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt loan_status  
0                5                 1       22800    low_risk  
1                3                 0       13600    low_risk  
2                3                 0       16100    low_risk  
3                5                 1       22700    low_risk  
4                5                 1       23000    low_risk

# Split the Data into Training and Testing

In [32]:
# Create our features
X = df.drop("loan_status", axis=1)
X = pd.get_dummies(X)
# Create our target
y = df['loan_status']

In [33]:
X.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  homeowner_mortgage  \
count     77536.000000      77536.000000  77536.000000        77536.000000   
mean          3.826610          0.392308  19221.949804            0.497472   
std           1.904426          0.582086   8371.635077            0.499997   
min           0.000000          0.000000      0.000000            0.000000   
25%           3.000000          0.000000  14800.000000            0.000000   
50%           4.000000          0.000000  18100.000000            0.000000   
75%           4.000000          1.000000  21400.000000            1.000000   
max          16.000000          3.000000  75200.000000            1.000000   

       homeowner_own  homeowner_rent  
count   77536.000000    77536.000000  
mean        0.398911        0.103616  
std         0.489678        0.304764  
min         0.000000        0.000000  
25%         0.000000        0.000000  
50%         0.000000        0.000000  
75%         1.000000        0.000000  
max         1.000000        1.000000

In [34]:
# Check the balance of our target values
y.value_counts()

low_risk     75036
high_risk     2500
Name: loan_status, dtype: int64

In [35]:
from sklearn.model_selection import train_test_split

In [36]:
# Create X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [37]:
X_train

loan_size  interest_rate  borrower_income  debt_to_income  \
29175     8600.0          6.792            44500        0.325843   
23020     7800.0          6.419            41000        0.268293   
31269    10000.0          7.386            50100        0.401198   
35479     9300.0          7.093            47300        0.365751   
13470     9200.0          7.045            46900        0.360341   
...          ...            ...              ...             ...   
20609     7200.0          6.177            38700        0.224806   
21440    10000.0          7.389            50100        0.401198   
73349    10200.0          7.463            50800        0.409449   
50057    11100.0          7.838            54400        0.448529   
5192     10600.0          7.632            52400        0.427481   

       num_of_accounts  derogatory_marks  total_debt  homeowner_mortgage  \
29175                3                 0       14500                   0   
23020                2                 0       11000                   0   
31269                4                 1       20100                   0   
35479                3                 0       17300                   0   
13470                3                 0       16900                   0   
...                ...               ...         ...                 ...   
20609                1                 0        8700                   0   
21440                4                 1       20100                   0   
73349                4                 1       20800                   0   
50057                5                 1       24400                   1   
5192                 5                 1       22400                   0   

       homeowner_own  homeowner_rent  
29175              1               0  
23020              1               0  
31269              1               0  
35479              0               1  
13470              1               0  
...              ...             ...  
20609              1               0  
21440              1               0  
73349              1               0  
50057              0               0  
5192               1               0  

[58152 rows x 10 columns]

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [38]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [39]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [40]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Simple Logistic Regression

In [41]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_train, y_train)

LogisticRegression(random_state=1)

In [42]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = model.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9520479254722232

In [43]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[  563,    56],
       [  102, 18663]], dtype=int64)

In [44]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.85      0.91      0.99      0.88      0.95      0.90       619
   low_risk       1.00      0.99      0.91      1.00      0.95      0.91     18765

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



# Oversampling

In this section, you will compare two oversampling algorithms to determine which algorithm results in the best performance. You will oversample the data using the naive random oversampling algorithm and the SMOTE algorithm. For each algorithm, be sure to complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Print the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

### Naive Random Oversampling

In [46]:
# Resample the training data with the RandomOversampler
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler(random_state=1)
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

# View the count of target classes with Counter
Counter(y_resampled)

Counter({'low_risk': 56271, 'high_risk': 56271})

In [47]:
# Train the Logistic Regression model using the resampled data
lr_model = LogisticRegression(solver='lbfgs', random_state=1)
lr_model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [49]:
# Calculated the balanced accuracy score
ros_y_pred = lr_model.predict(X_test)
balanced_accuracy_score(y_test, ros_y_pred)

0.9936781215845847

In [50]:
# Display the confusion matrix
confusion_matrix(y_test, ros_y_pred)

array([[  615,     4],
       [  116, 18649]], dtype=int64)

In [51]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, ros_y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.84      0.99      0.99      0.91      0.99      0.99       619
   low_risk       1.00      0.99      0.99      1.00      0.99      0.99     18765

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



### SMOTE Oversampling

In [53]:
# Resample the training data with SMOTE
from imblearn.over_sampling import SMOTE
X_smote_resampled, y_smote_resampled = SMOTE(random_state=1,sampling_strategy='auto').fit_resample( X_train, y_train)

# View the count of target classes with Counter
Counter(y_smote_resampled)

Counter({'low_risk': 56271, 'high_risk': 56271})

In [56]:
# Train the Logistic Regression model using the resampled data
smote_model = LogisticRegression(solver='lbfgs', random_state=1)
smote_model.fit(X_smote_resampled, y_smote_resampled)

LogisticRegression(random_state=1)

In [57]:
# Calculated the balanced accuracy score
smote_y_pred = smote_model.predict(X_test)
balanced_accuracy_score(y_test, smote_y_pred)

0.9936781215845847

In [58]:
# Display the confusion matrix
confusion_matrix(y_test, smote_y_pred)

array([[  615,     4],
       [  116, 18649]], dtype=int64)

In [59]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, smote_y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.84      0.99      0.99      0.91      0.99      0.99       619
   low_risk       1.00      0.99      0.99      1.00      0.99      0.99     18765

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



# Undersampling

In this section, you will test an undersampling algorithm to determine which algorithm results in the best performance compared to the oversampling algorithms above. You will undersample the data using the Cluster Centroids algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [60]:
# Resample the data using the ClusterCentroids resampler
from imblearn.under_sampling import ClusterCentroids
cc = ClusterCentroids(random_state=1)
X_cc_resampled, y_cc_resampled = cc.fit_resample(X_train, y_train)

# View the count of target classes with Counter
Counter(y_cc_resampled)

Counter({'high_risk': 1881, 'low_risk': 1881})

In [61]:
# Train the Logistic Regression model using the resampled data
cc_model = LogisticRegression(solver='lbfgs', random_state=1)
cc_model.fit(X_cc_resampled, y_cc_resampled)

LogisticRegression(random_state=1)

In [62]:
# Calculate the balanced accuracy score
cc_y_pred = cc_model.predict(X_test)
balanced_accuracy_score(y_test, cc_y_pred)

0.9865149130022852

In [63]:
# Display the confusion matrix
confusion_matrix(y_test, cc_y_pred)

array([[  606,    13],
       [  112, 18653]], dtype=int64)

In [64]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, cc_y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.84      0.98      0.99      0.91      0.99      0.97       619
   low_risk       1.00      0.99      0.98      1.00      0.99      0.97     18765

avg / total       0.99      0.99      0.98      0.99      0.99      0.97     19384



# Combination (Over and Under) Sampling

In this section, you will test a combination over- and under-sampling algorithm to determine if the algorithm results in the best performance compared to the other sampling algorithms above. You will resample the data using the SMOTEENN algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [66]:
# Resample the training data with SMOTEENN
from imblearn.combine import SMOTEENN
smote_enn = SMOTEENN(random_state=1)
X_enn_resampled, y_enn_resampled = smote_enn.fit_resample(X_train, y_train)

# View the count of target classes with Counter
Counter(y_enn_resampled)

Counter({'high_risk': 55622, 'low_risk': 55948})

In [67]:
# Train the Logistic Regression model using the resampled data
enn_model = LogisticRegression(solver='lbfgs', random_state=1)
enn_model.fit(X_enn_resampled, y_enn_resampled)

LogisticRegression(random_state=1)

In [68]:
# Calculate the balanced accuracy score
enn_y_pred = enn_model.predict(X_test)
balanced_accuracy_score(y_test, enn_y_pred)

0.9935182494822666

In [69]:
# Display the confusion matrix
confusion_matrix(y_test, enn_y_pred)

array([[  615,     4],
       [  122, 18643]], dtype=int64)

In [70]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, enn_y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.83      0.99      0.99      0.91      0.99      0.99       619
   low_risk       1.00      0.99      0.99      1.00      0.99      0.99     18765

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



# Final Questions

1. Which model had the best balanced accuracy score?

   The best balance accuracy score were the oversampling algorithims of the Naive Random Sample and the SMOTE models. The
   SMOTEENN (Combination) also did just as good.

2. Which model had the best recall score?

    Once again the oversampling algorithims (SMOTEENN and Naive Random Oversampling) had the highest recall score. 
    Although the SMOTEENN and Simple Logistic Regression had an overall score of .99, the high_risk target was only .98 and .91
    respectfully. 

3. Which model had the best geometric mean score?

    All models except for the Simple Logistic Regression model had a geometric mean score of .99 for both the high risk
    and low risk targets. 
